### PR TITLE
new nuget directory for building ANTLR4 C++ runtime as 3 Nuget packages

### PR DESCRIPTION
**Pull request by zosrothko**
Signed-off-by: Francis ANDRE <francis.andre.kampbell@orange.fr>

Hello

This PR is adding a directory antlr4\runtime\Cpp\runtime\nuget which contains a DOS script pack.cmd that is building the c++ antlr4-runtime as 3 Nuget packages.

usage: pack vsversion release [prerelease]
as for example :

    pack antlr4-runtime compiled with vs2019 for version 4.9.1 with prerelease '-rc': pack 2019 4.9.1 -rc
    pack antlr4-runtime compiled with vs2019 for version 4.9.1 : pack 2019 4.9.1

### CODE REVIEW

File: runtime/Cpp/runtime/nuget/pack.cmd [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+      if %VS_VERSION%==vs2019 (
+        if %PLATFORM%==x64 (
+          call "%VS160COMNTOOLS%%VS_VARSALL%" x86_amd64 8.1
+        ) else (
+          call "%VS160COMNTOOLS%%VS_VARSALL%" x86 8.1
+        )
+      )
+    )
+  )
+)
```
Comment (KvanTTT) [inline runtime/Cpp/runtime/nuget/pack.cmd:63]: Please drop support of VS < 2019 (2017, 2015) and add support of VS2022. It was updated in https://github.com/antlr/antlr4/pull/3725

**Comment by zosrothko** [comment]
vs2015 & vs2017 dropped. vs2022 added.

File: runtime/Cpp/runtime/nuget/pack.cmd [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+set PLATFORM=Win32
+
+rem -version ^^[16.0^^,17.0^^)
+set VS_VERSION=vs%1
+rem  should be set "VSWHERE='%ProgramFiles(x86)%\Microsoft Visual Studio\Installer\vswhere.exe  -property installationPath -version ^[16.0^,17.0^)'"
+if %VS_VERSION%==vs2019 (
+  set "VSWHERE='C:\PROGRA~2\"Microsoft Visual Studio"\Installer\vswhere.exe  -latest -property installationPath -version ^[16.0^,17.0^)'"
+) else (
+if %VS_VERSION%==vs2022 (
+  set "VSWHERE='C:\PROGRA~2\"Microsoft Visual Studio"\Installer\vswhere.exe  -latest -property installationPath -version ^[17.0^,18.0^)'"
```
Comment (KvanTTT) [inline runtime/Cpp/runtime/nuget/pack.cmd:20]: Does it work with vs2022? It's x64 and it's being installed to `Program Files` unlike previous versions that are being installed to `Program Files (x86)`.

File: runtime/Cpp/runtime/nuget/ANTLR4.Runtime.cpp.shared.nuspec [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+    <version>$version$$pre$</version>
+    <title>ANTLR4 Runtime c++ vs$vs$ shared</title>
+    <authors>Terence Parr &amp; Contributors</authors>
+    <requireLicenseAcceptance>true</requireLicenseAcceptance>
+    <license type='expression'>BSL-1.0</license>
+    <icon>image\antlr4.jpg</icon>
+    <projectUrl>https://www.antlr.org/</projectUrl>
+    <description>ANTLR (ANother Tool for Language Recognition) is a powerful parser generator for reading, processing, executing, or translating structured text or binary files. It's widely used to build languages, tools, and frameworks. From a grammar, ANTLR generates a parser that can build and walk parse trees..</description>
+    <releaseNotes>releaseNotes</releaseNotes>
+    <copyright>Copyright 2006-2020</copyright>
```
Comment (KvanTTT) [inline runtime/Cpp/runtime/nuget/ANTLR4.Runtime.cpp.shared.nuspec:14]: Probably the end year should be updated:

```suggestion
    <copyright>Copyright 2006-2022</copyright>
```

File: runtime/Cpp/runtime/nuget/ANTLR4.Runtime.cpp.noarch.nuspec [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+    <version>$version$$pre$</version>
+    <title>ANTLR4 Runtime c++ vs$vs$ $link$</title>
+    <authors>Terence Parr &amp; Contributors</authors>
+    <requireLicenseAcceptance>true</requireLicenseAcceptance>
+    <license type='expression'>BSL-1.0</license>
+    <icon>image\antlr4.jpg</icon>
+    <projectUrl>https://www.antlr.org/</projectUrl>
+    <description>ANTLR (ANother Tool for Language Recognition) is a powerful parser generator for reading, processing, executing, or translating structured text or binary files. It's widely used to build languages, tools, and frameworks. From a grammar, ANTLR generates a parser that can build and walk parse trees..</description>
+    <releaseNotes>See project URL</releaseNotes>
+    <copyright>Copyright 2014-2020</copyright>
```
Comment (KvanTTT) [inline runtime/Cpp/runtime/nuget/ANTLR4.Runtime.cpp.noarch.nuspec:14]:
```suggestion
    <copyright>Copyright 2014-2022</copyright>
```

File: runtime/Cpp/runtime/nuget/pack.cmd (hunk 2 above)
Comment (zosrothko) [reply runtime/Cpp/runtime/nuget/pack.cmd:20]: Yes. wswhere.exe is always in Program File(x86). You can test it by running
```
C:\PROGRA~2\"Microsoft Visual Studio"\Installer\vswhere.exe  -latest -property installationPath -version ^[17.0^,18.0^)
```

File: runtime/Cpp/runtime/nuget/ANTLR4.Runtime.cpp.shared.nuspec (hunk 3 above)
Comment (zosrothko) [reply runtime/Cpp/runtime/nuget/ANTLR4.Runtime.cpp.shared.nuspec]: Fixed

File: runtime/Cpp/runtime/nuget/ANTLR4.Runtime.cpp.noarch.nuspec (hunk 4 above)
Comment (zosrothko) [reply runtime/Cpp/runtime/nuget/ANTLR4.Runtime.cpp.noarch.nuspec]: Fixed

**Comment by zosrothko** [comment]
@KvanTTT Can this PR be applied please?

**Comment by KvanTTT** [comment]
It's up to @parrt

**Comment by zosrothko** [comment]
> Also, could you point out how and where nuget for C++ is used? I have never heard about that, I've heard only about .NET platform.

You can get the Nuget packages ANTLR4 Runtime for C++ here: https://int.nugettest.org/. Look for `ANTLR4.Runtime.vs2019` You can install those 3 packages in a MSVC C++ projet using the 'Nuget Package Manager' 

The PR is just adding to the source tree a pack.cmd that build those Nuget package for a new version or snapshot

**Comment by parrt** [comment]
I believe I manually upload to nuget from circlei builds or appveyor I think.

**Comment by parrt** [comment]
oh wow! i didn't know nuget did anything but c#!

**Pull request merged**